### ISSUE TITLE
[BUG] SIGSEGV after upgrading to v0.49.0 from v0.44.1

### ISSUE THREAD

**Issue by gornostal** [DataDog/dd-trace-php#1067]
### Bug description
I've started seeing this issue after the upgrade to 0.49.0. It occurs rarely and only in production.

I'd be happy to provide core dump but, there's something wrong with my php-fpm config (probably), because it doesn't generate core dump when this issue occurs. 
BTW, if you know reliable steps to get php to segfault, please let me know :)

At this point I'm looking for some guidance on what I can check to narrow down the potential causes.

Thanks!

### PHP version
```
PHP 7.4.11 (cli) (built: Oct 18 2020 19:44:48) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.11, Copyright (c), by Zend Technologies
    with ddtrace v0.49.0, Copyright Datadog, by Datadog
```
### Installed extensions
<details><summary>extensions</summary>
<p>

```
[PHP Modules]                                                                                                                                                                                                                                  
bcmath                                                                                                                                                                                                                                         
calendar                                                                                                                                                                                                                                       
Core                                                                                                                   
ctype                                          
curl                                                                                                                   
date                                                                                                                                                                                                                                           
ddtrace                                                    
dom                                                                                                                                                                                                                                            
exif                                                                                                                                                                                                                                           
FFI                                                                                                                    
fileinfo                                    
filter                                                                                                                 
ftp                                                                                                                    
gd                                                   
gettext                                                    
gmp                                                                                                                    
hash                                                                                                                                                                                                                                           
iconv                                                                                                                                                                                                                                          
igbinary                                                                                                               
intl                                                                                                                                                                                                                                           
json                                                                                                                   
libxml                               
mbstring                                                                                                               
mysqli                                                     
mysqlnd                                                  
openssl                                                    
pcntl                                                      
pcre                                                                                                                   
PDO                                                                                                                    
pdo_mysql                                                                                                                                                                                                                                      
Phar                                                                                                                                                                                                                                           
posix                                                                                                                  
rdkafka                                                                                                                
readline                                                                                                               
redis                                                                                                                  
Reflection                                           
session                                                                                                                
shmop                                                                                                                  
SimpleXML                                                                                                                                                                                                                                      
sockets                                                    
sodium                                                 
SPL                                                                                                                                                                                                                                            
ssh2                                                                                                                   
standard
sysvmsg
sysvsem
sysvshm
tokenizer
xdebug
xml
xmlreader
xmlwriter
xsl
Zend OPcache
zip
zlib

[Zend Modules]
Zend OPcache
ddtrace
```
</p>
</details>

### OS info
```
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
```

### Diagnostics and configuration

#### Output of phpinfo() (ddtrace >= 0.47.0)
![image](https://user-images.githubusercontent.com/1202543/97605007-ce704900-1a16-11eb-892b-6aac37a87b21.png)

<!-- If this issue is related to the CLI SAPI, copy the output of `php --ri=ddtrace` and paste it here. -->

### Upgrading info
Previous version 0.44.1

**Comment by labbati** [comment]
Hi @gornostal, thank you for the report. I have a few questions:

1. how to you install PHP? Maybe via the ondrej repo?
2. Are you setting `opcache.preload`?

For the core dump, I'd start verifying 3 things:
1. Create a folder for them and make sure the php-fpm user (assuming `www-data` has access to it). E.g. `mkdir -p /tmp/coredumps` and `chown -R /tmp/coredumps www-data:www-data`.
2. Tell your system to dump core dumps there: `echo '/tmp/coredumps/core' > /proc/sys/kernel/core_pattern`
3. Add `rlimit_core = unlimited` to the `[www]` (or whatever the pool is named) section of your php-fpm config file.

**Comment by labbati** [comment]
One more questions. Are you using only the out of the box automatic instrumentation or are you doing manual instrumentation? E.g. importing our package in composer?

**Comment by gornostal** [comment]
Thanks for the prompt reply @labbati!

* I'm installing PHP from here `"deb https://packages.sury.org/php/ $(lsb_release -sc) main"`
* No, I'm not using `opcache.preload`
* I'm using manual instrumentation too

> make sure the php-fpm user (assuming www-data) has access to it

That's probably what I missed in my config.
Do you know any reliable steps to cause any core dump by any chance? I want to verify that I can get core dumps locally before releasing this change to production.

**Comment by morrisonlevi** [comment]
You can use `posix_kill(posix_getpid(), $signal = 11)` to send a SIGSEGV to the currently running process, which should cause a core dump if everything is configured correctly.

**Comment by gornostal** [comment]
Thanks @morrisonlevi!

**Comment by labbati** [comment]
Thank you @gornostal. Last question, would you provide a couple of representative examples to what you do in manual instrumentation? Also are you importing the tracer from composer or relying on the classes provided by our extension (regardless of composer)?

**Comment by gornostal** [comment]
* Actually, we are not using manual instrumentation, if by that you mean creating new spans. In PHP code we use it for metrics + automatic instrumentation
* Importing from composer

**Comment by labbati** [comment]
Hi @gornostal, any success with the core dumps?

Thank you for the clarification about manual instrumentation. You are probably then referring to the [datadog/php-dadatadogstats](https://packagist.org/packages/datadog/php-datadogstatsd) package added to composer. In this case this has not impact with the this tracer.

**Comment by gornostal** [comment]
In composer.json I have two DD packages:
```
        "datadog/php-datadogstatsd": "^1.4",
        "datadog/dd-trace": "^0.49.0",
```
Also I install dd-php-tracer in Dockerfile like this:
```
curl https://github.com/DataDog/dd-trace-php/releases/download/0.49.0/datadog-php-tracer_0.49.0_amd64.deb -o /tmp/dd-tracer.deb -L && \
apt-get --yes install /tmp/dd-tracer.deb
```

As I learned more about core dumps, found that they are created on a host machine, not in Docker container (if everything is enabled correctly, of course). It will be tricky to get them from production cluster.

What I also noticed is that the crash doesn't affect website functionality nor it affects traces and metrics in DD UI.

**Comment by labbati** [comment]
Why do you add `datadog/dd-trace` then? If you are not doing any manual instrumentation (meaning you do not use classes in the `use DDTrace\...` namespace, you can omit it. Otherwise can you share snippets of code you are using?

**Comment by labbati** [comment]
Hi @gornostal thanks for the snippet. We are about to try to investigate code paths triggered by that.

> As I learned more about core dumps, found that they are created on a host machine, not in Docker container (if everything is enabled correctly, of course). It will be tricky to get them from production cluster.

Actually it is possible to collect them in docker as well. If you tell us on which docker image you are, we can help providing a sample `Dockerfile` illustrating how to do that.

> It occurs rarely and only in production.

Any chance you are able to reproduce this issue in a staging/testing container putting it under some load? As we proceed with the investigation I might be required to ask you to test some solutions and I'd prefer to avoid messing with your production.

> What I also noticed is that the crash doesn't affect website functionality nor it affects traces and metrics in DD UI.

I find it to be very strange. If there is a seg fault the php-fpm process is killed, so there must be some effect on the page rendered, unless it happens after everything is sent (headers and content). This can point us in the direction of our request shutdown hook that we use to flush spans produced in the context of the current request, but I need to think a little bit more about it.

**Comment by gornostal** [comment]
> If you tell us on which docker image you are

We are using `debian:buster` as a base image.

> Any chance you are able to reproduce this issue in a staging/testing container putting it under some load?

No, not yet. I'm working on this now.

> What I also noticed is that the crash doesn't affect website functionality nor it affects traces and metrics in DD UI.

Please disregard that comment. I just meant that no user has complained about errors (probably because 1 out of 3k requests fails in SIGSEGV) and I couldn't find any more application errors that could be related to a crash.

**Comment by gornostal** [comment]
Just FYI, I haven't been able to reproduce the issue with load testing yet.

However, I have another issue that I report separately.

**Comment by labbati** [comment]
Hi @gornostal. I am reaching out to know if you had any more time to spend on this issue.

As a summary of where we are:
- we tried to load test our testing infrastructure with your docker image/helper class, etc...
- we are not able to recreate a SEGFAULT, even fuzzing things that are risky such as opcache configurations

As of now and without any further information, I think that we are shouting in the dark. I propose for us to take one of the following two approaches to move forward:

1. We can have a pair programming session and I can help disabling various integrations/features and enabling them one by one to at least try to spot the offending 'area'.
    - this would require to have access to an environment, preferably staging/testing, where we can do frequent restarts of the php-fpm process.
    - this approach does make sense if we are able to call a state as broken or working in minutes (i.e. if the SEGFAULT typically happens every few minutes at most).
2. We try to recreate it enabling core dumps. In this case we can still have a pairing session so I can help with core dumps configuration. Even if that depends on the permissions that your docker engine has.

Please let us know what if any of the plans above works for you.

**Comment by gornostal** [comment]
Hi @labbati
Thanks for checking on this. I have a few updates actually.

Unfortunately we were not able to reproduce it in environments other than prod. In prod we tried to enable core dumps but attempts were unsuccessful -- apparently due to ulimits that are set on host OS that we have no access to. We also decided it's not worth spending our time on this issue because after we downgraded the library installed via `apt` and `composer` to v0.42.2 we see crashes only once every two hours, which is something we can live with for now.

We still see a lot of verbose logging #1073 from curl, but I think I'm going to be able to find a cause of that issue soon.

**Comment by labbati** [comment]
Hi @gornostal thank you for the reply. I understand that proceeding with the investigation would require some time and that you are busy. I am very grateful for the time you invested on this so far. Thank you.

As a last note, if you picked 0.45.2 because is the most recent that presents the issue more rarely, I have to ask one last question, since that version provided a radical change in the way we load our PHP code. Any chance you are using PHP inis `auto_prepend_file` and `auto_append_file`?

**Comment by gornostal** [comment]
No, we're not using `auto_prepend_file` or `auto_append_file`.

BTW, my gut feel is the Redis integration could be the cause. We use Redis pretty heavily. Although I haven't tried to use 0.49.0 with disabled Redis integration to verify that hypothesis.

**Comment by labbati** [comment]
@gornostal I am closing this issue since we have not a way to proceed with the investigation without a core dump. Please feel free to reach out to support in case you are interested with moving forward with the investigation and together we can agree on a plan to move forward. Thank you for your patience.

**Comment by gornostal** [comment]
No problem.
Thanks for your help with this!
